### PR TITLE
Strongly typed value serialization and deserialization using IJSValueReader, JSValue, and IJSValueWriter

### PR DESCRIPTION
**Pull request by vmoroz**
While we work on the new RNW native module and view managers implementation, we need an efficient way to send data through the ABI. The core of the data transfer implementation are the IJSValueReader and IJSValueWriter interfaces. These interfaces allow the most efficient serialized data transfer that reduces requirements for memory allocations. E.g if we would use boxing and transfer all values as IInspectable, then for any numeric values we would require to allocate a ref counted object that implements the IInspectable.

The IJSValueReader and IJSValueWriter interfaces were introduced earlier as a part of the new RNW native module infrastructure. We were missing good examples how to use and implement them.
While working on the examples, we realized that we need to do more than just set of samples. We had to re-define them and build additional infrastructure to simplify their usage.

Since reading complex objects from IJSValueReader can be a challenge in some complex cases we have introduced JSValue which is very similar to folly::dynamic, but immutable. While it is easier to use JSValue to read data, it requires more resources than IJSValueReader. But combining these two methods together, developers can address any complex situations.

In C# to implement reading custom value type from IJSValueReader developers should add ReadValue extension for IJSValueReader or JSValue types. They can use IJSValueReader.ReadValue<T>() to call the ReadValue. This method will call either one of these custom extensions, or one of the predefined extensions. The predefined extensions are offered for simple standard types, enums, nullables, dictionaries, collections, tuples, variants (OneOf), JSValue, and simple structs. To write custom value type to IJSValueWriter, developers should add WriteValue extension for IJSValueWriter. We offer the ame set of predefined WriteValue extensions.

In C++ developers could do the same, but instead of creating extension methods, they should add ReadValue and WriteValue function overloads.

The new unit tests for C++ and C# that show how create such methods.

In this change we also have changed the way we register and invoke native modules to reflect the new value serialization. In C++ the REACT_* macros that used for native module member registration are changed to be simple static functions. It helps to eliminate the need to use an instance Boolean field per registered member, and we also do the registration 'on-demand' after the static initialization. 

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-windows/pull/3760)

### CODE REVIEW

**Comment by NickGerleman** [comment]
@vmoroz do you know what's going on here? 
image

**Comment by vmoroz** [comment]
@ngerlem@microsoft.com this change has a lot of new code including two new unit test projects. I do not know how GitHub counts images that are part of the new test project.

---
In reply to: [564710850](https://github.com/microsoft/react-native-windows/pull/3760#issuecomment-564710850) [](ancestors = 564710850)

File: packages/microsoft-reactnative-sampleapps/windows/SampleLibraryCPP/CircleViewManagerCPP.cpp [hunk 1]
```diff
@@ -38,7 +38,7 @@ FrameworkElement CircleViewManagerCPP::CreateView() noexcept {
 
 // IViewManagerWithChildren
 
-void CircleViewManagerCPP::AddView(FrameworkElement const &parent, UIElement const &child, int64_t index) noexcept {
+void CircleViewManagerCPP::AddView(FrameworkElement const &parent, UIElement const &child, int64_t /*index*/) noexcept {
```
Comment (NickGerleman) [inline packages/microsoft-reactnative-sampleapps/windows/SampleLibraryCPP/CircleViewManagerCPP.cpp]: Did you add this change as part of this PR? If so, could we separate the unrelated changes out? If not I think your branch might be in a weird state.

File: vnext/Microsoft.ReactNative.SharedManaged/ReactMethodInfo.cs [hunk 2]
```diff
@@ -2,304 +2,204 @@
...
-      ParameterExpression moduleParameter = Expression.Parameter(typeof(object), "module");
-      ParameterExpression inputReaderParameter = Expression.Parameter(typeof(IJSValueReader), "inputReader");
-      ParameterExpression outputWriterParameter = Expression.Parameter(typeof(IJSValueWriter), "outputWriter");
-      ParameterExpression resolveParameter = Expression.Parameter(typeof(MethodResultCallback), "resolve");
-      ParameterExpression rejectParameter = Expression.Parameter(typeof(MethodResultCallback), "reject");
-
-      // Input variables to read from inputReader
-      ParameterExpression[] inputVariables = new ParameterExpression[parameters.Length - 1];
-      for (int i = 0; i < inputVariables.Length; ++i)
+      else if (isPromise(lastParameterType))
```
Comment (NickGerleman) [inline vnext/Microsoft.ReactNative.SharedManaged/ReactMethodInfo.cs]: >lastParameterType [](start = 25, length = 17)

Can we have the promise be the return type instead of the last parameter?

File: vnext/Microsoft.ReactNative.SharedManaged/ReactMethodInfo.cs [hunk 3]
```diff
@@ -2,304 +2,204 @@
...
-      {
-        MethodReturnType = MethodReturnType.Callback;
-        MethodImpl = new Lazy<ReactMethodImpl>(() => MakeReturningMethod(methodInfo, returnType, parameters), LazyThreadSafetyMode.PublicationOnly);
-      }
-      else
+      Type lastParameterType = parameters.Length > 0 ? parameters[parameters.Length - 1].ParameterType : null;
+      Type secondToLastParameterType = parameters.Length > 1 ? parameters[parameters.Length - 2].ParameterType : null;
+      Func<ReactMethodImpl> createMethod;
+
+      bool isPromise(Type type)
```
Comment (NickGerleman) [inline vnext/Microsoft.ReactNative.SharedManaged/ReactMethodInfo.cs]: >isPromise [](start = 11, length = 9)

Would this be cleaner as a private method on the class instead of a method inside of a method?

File: packages/microsoft-reactnative-sampleapps/windows/SampleLibraryCPP/CircleViewManagerCPP.cpp (hunk 1 above)
Comment (vmoroz) [reply packages/microsoft-reactnative-sampleapps/windows/SampleLibraryCPP/CircleViewManagerCPP.cpp]: I have eliminated a warning caused by the unused variable. My branch is in a good state.

File: vnext/Microsoft.ReactNative.SharedManaged/ReactMethodInfo.cs (hunk 2 above)
Comment (NickGerleman) [inline vnext/Microsoft.ReactNative.SharedManaged/ReactMethodInfo.cs]: >lastParameterType [](start = 25, length = 17)

Having a promise as an explicit parameter feels a bit weird to me. How difficult would it be to be able to use stock c# functions returning something like `Task<ReactResult<T>>`? That would have the benefit of forcing methods to return something.

File: vnext/Microsoft.ReactNative.Cxx/Crash.h [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+#define VerifyElseCrash(condition) \
+  do {                             \
+    if (!(condition)) {            \
+      assert(false && #condition); \
+      std::terminate();            \
+    }                              \
+  } while (false)
+
+#define VerifyElseCrashSz(condition, message) \
```
Comment (NickGerleman) [inline vnext/Microsoft.ReactNative.Cxx/Crash.h]: >Sz [](start = 23, length = 2)

Not a fan of adding something with Hungrian notation to RNW, even if it matches what we have in devmain.

**Comment by NickGerleman** [comment]
>  *  Generated: 2019-11-15 15:01:56.628356

Doesn't RNW vnext already use Google Test? Why are we adding another UT framework?

---
Refers to: vnext/Microsoft.ReactNative.Cxx.UnitTests/catch.hpp:3 in d7599a1. [](commit_id = d7599a1fbbb88300ccd2b3e280b74fda9f41f061, deletion_comment = False)

File: packages/microsoft-reactnative-sampleapps/index.windows.js [hunk 5]
```diff
@@ -38,6 +38,12 @@ var getCallback = function(prefix) {
   };
 };
 
+var getErrorCallback = function(prefix) {
+  return function(error) {
+    log(prefix + (error || {}).message);
```
Comment (NickGerleman) [inline packages/microsoft-reactnative-sampleapps/index.windows.js]: >|| {} [](start = 24, length = 5)

What is the motivation for this change?

File: vnext/Microsoft.ReactNative.Cxx/ReactPromise.h [hunk 6]
```diff
@@ -0,0 +1,56 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+#include "ReactError.h"
+#include "winrt/Microsoft.ReactNative.Bridge.h"
+
+namespace winrt::Microsoft::ReactNative::Bridge {
```
Comment (NickGerleman) [inline vnext/Microsoft.ReactNative.Cxx/ReactPromise.h]: >winrt [](start = 10, length = 5)

Do we use this namespace for things outside of cppwinrt? I immediately thought this type was part of that based on namespace when I saw it's first usage,

File: vnext/Microsoft.ReactNative.Cxx/ReactError.h [hunk 7]
```diff
@@ -0,0 +1,15 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+#include "JSValue.h"
+
+namespace winrt::Microsoft::ReactNative::Bridge {
+
+struct ReactError {
```
Comment (NickGerleman) [inline vnext/Microsoft.ReactNative.Cxx/ReactError.h]: >ReactError [](start = 7, length = 10)

Could we document the fields here? E.g. I wouldn't be sure looking at this as to what "Code" or "UserInfo" represents.

File: vnext/Microsoft.ReactNative.SharedManaged/ReactMethodInfo.cs (hunk 2 above)
Comment (vmoroz) [reply vnext/Microsoft.ReactNative.SharedManaged/ReactMethodInfo.cs]: Good idea! We can consider adding it in future. We use similar approach in our Reka implementation by returning Mso::Future. Though, I find the creation and managing the return result a little bit more complex in that case. While Task<T> enforces developers to return something, please note that developers write this method themselves and returning Task<T> instead of receiving IReactPromise<T> would also require them to create it and set instead of just setting values. Plus the whole business of scheduling task on another thread by default can also end up to be a non-trivial task.

File: vnext/Microsoft.ReactNative.Cxx/Crash.h (hunk 4 above)
Comment (vmoroz) [reply vnext/Microsoft.ReactNative.Cxx/Crash.h]: Any proposal for a different name?

File: packages/microsoft-reactnative-sampleapps/index.windows.js (hunk 5 above)
Comment (vmoroz) [reply packages/microsoft-reactnative-sampleapps/index.windows.js]: The error object has a member 'message' the ||{} technique is used to ensure that we do not crash should the error be null.

File: vnext/Microsoft.ReactNative.Cxx/ReactPromise.h (hunk 6 above)
Comment (vmoroz) [reply vnext/Microsoft.ReactNative.Cxx/ReactPromise.h]: I want to address the whole issue about namespaces as a separate PR. At this point it was just easier to keep the same namespace to avoid long names.

File: vnext/Microsoft.ReactNative.SharedManaged/ReactMethodInfo.cs (hunk 3 above)
Comment (vmoroz) [reply vnext/Microsoft.ReactNative.SharedManaged/ReactMethodInfo.cs]: Maybe. I can revisit it later.

---
In reply to: [356823803](https://github.com/microsoft/react-native-windows/pull/3760#discussion_r356823803) [](ancestors = 356823803)

**Comment by vmoroz** [comment]
>  *  Generated: 2019-11-15 15:01:56.628356

It was the only framework that was recommended to be used to test C++/WinRT code. We can remove/replace it if we find that Google Test or Microsoft test can be used for it too.  My main focus in this PR was to implement serialization/deserialization and all my attempts to use other unit test frameworks were not successful yet. 

---
In reply to: [564728425](https://github.com/microsoft/react-native-windows/pull/3760#issuecomment-564728425) [](ancestors = 564728425)

---
Refers to: vnext/Microsoft.ReactNative.Cxx.UnitTests/catch.hpp:3 in d7599a1. [](commit_id = d7599a1fbbb88300ccd2b3e280b74fda9f41f061, deletion_comment = False)

**Comment by NickGerleman** [comment]
>  *  Generated: 2019-11-15 15:01:56.628356

I would be interested in hearing more here. What were the incompatibilities you were concerned about?

---
In reply to: [564767446](https://github.com/microsoft/react-native-windows/pull/3760#issuecomment-564767446) [](ancestors = 564767446,564728425)

---
Refers to: vnext/Microsoft.ReactNative.Cxx.UnitTests/catch.hpp:3 in d7599a1. [](commit_id = d7599a1fbbb88300ccd2b3e280b74fda9f41f061, deletion_comment = False)

File: vnext/Microsoft.ReactNative.Cxx/ReactPromise.h (hunk 6 above)
Comment (NickGerleman) [reply vnext/Microsoft.ReactNative.Cxx/ReactPromise.h]: Could you add a Github issue for this to RNW with the "JSHost" tag?

---
In reply to: [356845271](https://github.com/microsoft/react-native-windows/pull/3760#discussion_r356845271) [](ancestors = 356845271)

File: vnext/Microsoft.ReactNative.Cxx/Crash.h (hunk 4 above)
Comment (NickGerleman) [reply vnext/Microsoft.ReactNative.Cxx/Crash.h]: VerifyElseCrashMessage?

We also might have something in Folly that does this that might be worth reusing.

---
In reply to: [356831337](https://github.com/microsoft/react-native-windows/pull/3760#discussion_r356831337) [](ancestors = 356831337)

File: vnext/Microsoft.ReactNative.Cxx/JSValue.h [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+
+  ~JSValue() noexcept;
+
+  JSValue Copy() const noexcept;
+  static JSValueObject CopyObject(const JSValueObject &other) noexcept;
+  static JSValueArray CopyArray(const JSValueArray &other) noexcept;
+
+  JSValueType Type() const noexcept;
+  bool IsNull() const noexcept;
+  const JSValueObject &Object() const noexcept;
```
Comment (NikoAri) [inline vnext/Microsoft.ReactNative.Cxx/JSValue.h]: >&Object() [](start = 22, length = 9)

This looks strange, but I guess it's required by formatting?

**Comment by NikoAri** [comment]
Why are we adding all empty png files for different scales???
Square150x150Logo.scale-200.png etc

File: vnext/Microsoft.ReactNative.Managed.UnitTests/Properties/AssemblyInfo.cs [hunk 9]
```diff
@@ -0,0 +1,18 @@
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+[assembly: AssemblyTitle("Microsoft.ReactNative.Managed.UnitTests")]
+[assembly: AssemblyDescription("")]
+[assembly: AssemblyConfiguration("")]
+[assembly: AssemblyCompany("")]
+[assembly: AssemblyProduct("Microsoft.ReactNative.Managed.UnitTests")]
+[assembly: AssemblyCopyright("Copyright ©  2019")]
```
Comment (NikoAri) [inline vnext/Microsoft.ReactNative.Managed.UnitTests/Properties/AssemblyInfo.cs]: >   [](start = 41, length = 2)

(nit) remove extra space

File: vnext/Microsoft.ReactNative.Managed.UnitTests/Properties/AssemblyInfo.cs [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+[assembly: AssemblyTrademark("")]
+[assembly: AssemblyCulture("")]
+[assembly: AssemblyMetadata("TargetPlatform","UAP")]
+
+// [assembly: AssemblyVersion("1.0.*")]
```
Comment (NikoAri) [inline vnext/Microsoft.ReactNative.Managed.UnitTests/Properties/AssemblyInfo.cs]: >AssemblyVersion [](start = 14, length = 15)

remove?

File: vnext/Microsoft.ReactNative.Managed.UnitTests/Properties/AssemblyInfo.cs [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+using System.Runtime.InteropServices;
+
+[assembly: AssemblyTitle("Microsoft.ReactNative.Managed.UnitTests")]
+[assembly: AssemblyDescription("")]
+[assembly: AssemblyConfiguration("")]
+[assembly: AssemblyCompany("")]
+[assembly: AssemblyProduct("Microsoft.ReactNative.Managed.UnitTests")]
+[assembly: AssemblyCopyright("Copyright ©  2019")]
+[assembly: AssemblyTrademark("")]
+[assembly: AssemblyCulture("")]
```
Comment (NikoAri) [inline vnext/Microsoft.ReactNative.Managed.UnitTests/Properties/AssemblyInfo.cs]: >emblyCulture("")] [](start = 14, length = 17)

Can we just remove all empty fields?

File: vnext/Microsoft.ReactNative.Cxx/Crash.h (hunk 4 above)
Comment (NikoAri) [reply vnext/Microsoft.ReactNative.Cxx/Crash.h]: Would be nice to grab it from Open Source Mso :)

---
In reply to: [356901187](https://github.com/microsoft/react-native-windows/pull/3760#discussion_r356901187) [](ancestors = 356901187,356831337)

**Comment by vmoroz** [comment]
>  *  Generated: 2019-11-15 15:01:56.628356

Yes, we can discuss it offline. The starting point is that there is no C++/WinRT unit test project and we have to do something custom. I tried to bring MSTest to a C++/WinRT blank application and after a few hours I could not make it work. Let me know if you know about better alternative.

---
In reply to: [564788592](https://github.com/microsoft/react-native-windows/pull/3760#issuecomment-564788592) [](ancestors = 564788592,564767446,564728425)

---
Refers to: vnext/Microsoft.ReactNative.Cxx.UnitTests/catch.hpp:3 in d7599a1. [](commit_id = d7599a1fbbb88300ccd2b3e280b74fda9f41f061, deletion_comment = False)

**Comment by vmoroz** [comment]
This is part of UWP projects that have UI. These are all generated files and are need for sample/test projects to run.

---
In reply to: [564800474](https://github.com/microsoft/react-native-windows/pull/3760#issuecomment-564800474) [](ancestors = 564800474)

File: vnext/Microsoft.ReactNative/DynamicReader.cpp [hunk 12]
```diff
@@ -3,180 +3,144 @@
...
+    case folly::dynamic::Type::DOUBLE:
+      return JSValueType::Double;
+    case folly::dynamic::Type::INT64:
+      return JSValueType::Int64;
+    case folly::dynamic::Type::OBJECT:
+      return JSValueType::Object;
+    case folly::dynamic::Type::STRING:
+      return JSValueType::String;
+    default:
+      return JSValueType::Null;
```
Comment (NickGerleman) [inline vnext/Microsoft.ReactNative/DynamicReader.cpp]: >JSValueType::Null; [](start = 13, length = 18)

why this vs VEC?

File: vnext/Microsoft.ReactNative.Cxx/Crash.h (hunk 4 above)
Comment (vmoroz) [reply vnext/Microsoft.ReactNative.Cxx/Crash.h]: Yes, but I just needed something to crash conditionally for now. 

---
In reply to: [356912397](https://github.com/microsoft/react-native-windows/pull/3760#discussion_r356912397) [](ancestors = 356912397,356901187,356831337)

File: vnext/Microsoft.ReactNative.Cxx/JSValue.h [hunk 13]
```diff
@@ -0,0 +1,252 @@
...
+
+ private: // Instance fields
+  JSValueType m_type;
+  union {
+    JSValueObject m_object;
+    JSValueArray m_array;
+    std::string m_string;
+    bool m_bool;
+    int64_t m_int64;
+    double m_double;
```
Comment (NickGerleman) [inline vnext/Microsoft.ReactNative.Cxx/JSValue.h]: Seems like we kind of just reimplemented std::variant. Using variant might make some other things easier, e.g. I'm pretty sure the default move constructor would work instead of switching on type like we do right now.

File: vnext/Microsoft.ReactNative.Cxx/JSValue.h (hunk 8 above)
Comment (NickGerleman) [reply vnext/Microsoft.ReactNative.Cxx/JSValue.h]: Same for the custom dtor stuff we need to do

---
In reply to: [356910886](https://github.com/microsoft/react-native-windows/pull/3760#discussion_r356910886) [](ancestors = 356910886)

File: vnext/Microsoft.ReactNative.Cxx/JSValue.cpp [hunk 14]
```diff
@@ -0,0 +1,324 @@
...
+  }
+
+  return items;
+}
+
+//===========================================================================
+// Standalone inline functions implementation
+//===========================================================================
+
+void swap(JSValue &left, JSValue &right) noexcept {
```
Comment (NickGerleman) [inline vnext/Microsoft.ReactNative.Cxx/JSValue.cpp]: >swap [](start = 5, length = 4)

Does std::swap not do the right thing?

File: vnext/Microsoft.ReactNative.Cxx/JSValueTreeWriter.h [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+#ifndef MICROSOFT_REACTNATIVE_JSVALUETREEWRITER
+#define MICROSOFT_REACTNATIVE_JSVALUETREEWRITER
+
+#include <stack>
+#include "JSValue.h"
+
+namespace winrt::Microsoft::ReactNative::Bridge {
+
+// Writes to a tree of JSValue objects.
+struct JSValueTreeWriter : implements<JSValueTreeWriter, IJSValueWriter> {
```
Comment (NickGerleman) [inline vnext/Microsoft.ReactNative.Cxx/JSValueTreeWriter.h]: This is a bit general, but what are the main benefits we see of this compared to using Dynamic and DynamicConverter? Eithout context, it feels like there's some overlap.

File: vnext/Microsoft.ReactNative.Cxx/JSValueTreeWriter.h [hunk 16]
```diff
@@ -0,0 +1,51 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+#ifndef MICROSOFT_REACTNATIVE_JSVALUETREEWRITER
```
Comment (NickGerleman) [inline vnext/Microsoft.ReactNative.Cxx/JSValueTreeWriter.h]: >MICROSOFT_REACTNATIVE_JSVALUETREEWRITER [](start = 8, length = 39)

Why do we need a header guard if we have `#pragma once`?

File: vnext/Microsoft.ReactNative.Cxx/JSValue.h (hunk 8 above)
Comment (vmoroz) [reply vnext/Microsoft.ReactNative.Cxx/JSValue.h]: Yes, this is the formatting adopted by RNW project.
@NickGerleman, how is it related to the custom dtor?

---
In reply to: [356917132](https://github.com/microsoft/react-native-windows/pull/3760#discussion_r356917132) [](ancestors = 356917132,356910886)

File: vnext/Microsoft.ReactNative.Managed.UnitTests/Properties/AssemblyInfo.cs (hunk 11 above)
Comment (vmoroz) [reply vnext/Microsoft.ReactNative.Managed.UnitTests/Properties/AssemblyInfo.cs]: This is an auto-generated file. I did not pay much attention to it. I will revisit and clean it up at some point.

---
In reply to: [356912075](https://github.com/microsoft/react-native-windows/pull/3760#discussion_r356912075) [](ancestors = 356912075)

File: vnext/Microsoft.ReactNative.Cxx/JSValue.h (hunk 8 above)
Comment (NickGerleman) [reply vnext/Microsoft.ReactNative.Cxx/JSValue.h]: Whoops, I added that to the wrong thread, That was meant to be on the std::variant comment

---
In reply to: [356918726](https://github.com/microsoft/react-native-windows/pull/3760#discussion_r356918726) [](ancestors = 356918726,356917132,356910886)

File: vnext/Microsoft.ReactNative.Cxx/NativeModules.h [hunk 17]
```diff
@@ -453,19 +639,103 @@ struct ModuleEventFieldInfo<TFunc<void(TArg)> TModule::*> {
...
+ private:
+  void *m_module;
+  IReactModuleBuilder m_moduleBuilder;
+};
+
+template <class T>
+struct BoxedValue : implements<BoxedValue<T>, IBoxedValue> {
+  BoxedValue() noexcept {}
+
+  int64_t GetPtr() noexcept {
```
Comment (NickGerleman) [inline vnext/Microsoft.ReactNative.Cxx/NativeModules.h]: >int64_t [](start = 2, length = 7)

Is there a way to specify in IDL to use ptr width types (e.g. intptr_t). Using int64_t for pointers can get annoying if you ever need to do assignment on x86 since the compiler will worry about truncation.

File: vnext/Microsoft.ReactNative.Cxx/JSValue.h (hunk 13 above)
Comment (vmoroz) [reply vnext/Microsoft.ReactNative.Cxx/JSValue.h]: It is more compact than use of std::variant. I do not see issues with implementing move constructor. The std::variant must have done it too, and we would not have any saving on code size. The std::variant seems to me as an overkill for the case when we can just use union.

---
In reply to: [356916675](https://github.com/microsoft/react-native-windows/pull/3760#discussion_r356916675) [](ancestors = 356916675)

File: vnext/Microsoft.ReactNative.Cxx/ReactMemberInfo.h [hunk 18]
```diff
@@ -0,0 +1,17 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+#ifndef MICROSOFT_REACTNATIVE_REACTMEMBERINFO
```
Comment (NickGerleman) [inline vnext/Microsoft.ReactNative.Cxx/ReactMemberInfo.h]: Mentioned this on another instance of this, but do we need the header guard?

File: vnext/Microsoft.ReactNative.Cxx/ReactPromise.cpp [hunk 19]
```diff
@@ -0,0 +1,94 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "pch.h"
+#include "ReactPromise.h"
+#include "JSValueWriter.h"
+
+namespace winrt::Microsoft::ReactNative::Bridge {
+
+static const char *ErrorDefaultCode = "EUNSPECIFIED";
```
Comment (NickGerleman) [inline vnext/Microsoft.ReactNative.Cxx/ReactPromise.cpp]: >ErrorDefaultCode [](start = 19, length = 16)

nit: To me these seem more readbable with words reordered. E.g. `DefaultErrorCode`

File: vnext/Microsoft.ReactNative.Cxx/ReactMemberInfo.h (hunk 18 above)
Comment (NickGerleman) [reply vnext/Microsoft.ReactNative.Cxx/ReactMemberInfo.h]: Seeing this in a few more files as well

---
In reply to: [356920831](https://github.com/microsoft/react-native-windows/pull/3760#discussion_r356920831) [](ancestors = 356920831)

File: vnext/Microsoft.ReactNative.Cxx/JSValue.cpp (hunk 14 above)
Comment (vmoroz) [reply vnext/Microsoft.ReactNative.Cxx/JSValue.cpp]: Good point! It is the same as std::swap. I will remove it.

---
In reply to: [356917388](https://github.com/microsoft/react-native-windows/pull/3760#discussion_r356917388) [](ancestors = 356917388)

File: vnext/Microsoft.ReactNative.Cxx.UnitTests/JsonReader.cpp [hunk 20]
```diff
@@ -0,0 +1,379 @@
...
+//   Container: represents a [] or {}
+//   isStringContent: is the current value a string? Either a name or quoted value
+
+#include "pch.h"
+#include "JsonReader.h"
+#include "Crash.h"
+
+namespace winrt::Microsoft::ReactNative::Bridge {
+
+static bool IsSimpleWhitespace(wchar_t c) noexcept {
```
Comment (NickGerleman) [inline vnext/Microsoft.ReactNative.Cxx.UnitTests/JsonReader.cpp]: >IsSimpleWhitespace [](start = 12, length = 18)

This seems really similar to std::isspace

File: vnext/Microsoft.ReactNative.Cxx.UnitTests/JsonJSValueReader.cpp [hunk 21]
```diff
@@ -0,0 +1,160 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "pch.h"
+#include "JsonJSValueReader.h"
+
+namespace winrt::Microsoft::ReactNative::Bridge {
+
+//===========================================================================
+// JsonJSValueReader implementation
```
Comment (NickGerleman) [inline vnext/Microsoft.ReactNative.Cxx.UnitTests/JsonJSValueReader.cpp]: >JsonJSValueReader [](start = 3, length = 17)

Would this ever be useful outside of UT code? Wondering if we should share it outside the UT directory.

File: vnext/Microsoft.ReactNative.Cxx/JSValueTreeWriter.h (hunk 15 above)
Comment (vmoroz) [reply vnext/Microsoft.ReactNative.Cxx/JSValueTreeWriter.h]: folly::dynamic is not ABI safe. We cannot use it in the API. But developers want to have something like folly-dynamic is some cases. The JSValue is an immutable class to fulfill that need. We could use the folly::dynamic, but I want the native modules code that is outside of ABI boundary to not have any dependencies on any big third-party libraries. The Microsoft.ReactNative.Cxx is a small shared code library that we require to depend on.

---
In reply to: [356918350](https://github.com/microsoft/react-native-windows/pull/3760#discussion_r356918350) [](ancestors = 356918350)

File: vnext/Microsoft.ReactNative.Cxx/JSValueTreeWriter.h (hunk 16 above)
Comment (vmoroz) [reply vnext/Microsoft.ReactNative.Cxx/JSValueTreeWriter.h]: Pragma once only protects from multiple use of code if it came from the same file. If the same file came from different source locations due to symlinking or because it was copied, then it will not protect. It is generally safer to provide the guard macros in all shared header files.
The gurad does not replace "pragma once" because the pragma enables compiler not even open the file second time.

---
In reply to: [356918446](https://github.com/microsoft/react-native-windows/pull/3760#discussion_r356918446) [](ancestors = 356918446)

File: vnext/Microsoft.ReactNative.Cxx/NativeModules.h (hunk 17 above)
Comment (vmoroz) [reply vnext/Microsoft.ReactNative.Cxx/NativeModules.h]: Agee. But, no, the new IDL 3.0 does not allow use of pointers or integers of pointer size. [https://docs.microsoft.com/en-us/uwp/midl-3/intro](https://docs.microsoft.com/en-us/uwp/midl-3/intro)
It is very similar to C# in that sense.

---
In reply to: [356920177](https://github.com/microsoft/react-native-windows/pull/3760#discussion_r356920177) [](ancestors = 356920177)

File: vnext/Microsoft.ReactNative.Cxx/ReactMemberInfo.h (hunk 18 above)
Comment (vmoroz) [reply vnext/Microsoft.ReactNative.Cxx/ReactMemberInfo.h]: See my answer in other place.

---
In reply to: [356921647](https://github.com/microsoft/react-native-windows/pull/3760#discussion_r356921647) [](ancestors = 356921647,356920831)

File: vnext/Microsoft.ReactNative.Cxx/ReactPromise.cpp (hunk 19 above)
Comment (vmoroz) [reply vnext/Microsoft.ReactNative.Cxx/ReactPromise.cpp]: I will think about it.

---
In reply to: [356921223](https://github.com/microsoft/react-native-windows/pull/3760#discussion_r356921223) [](ancestors = 356921223)

File: vnext/Microsoft.ReactNative.Cxx.UnitTests/JsonReader.cpp (hunk 20 above)
Comment (vmoroz) [reply vnext/Microsoft.ReactNative.Cxx.UnitTests/JsonReader.cpp]: It maybe similar, but not the same. We must follow here the JSON specification. It states that it must be only these four characters: [https://www.json.org/json-en.html](https://www.json.org/json-en.html)

---
In reply to: [356922292](https://github.com/microsoft/react-native-windows/pull/3760#discussion_r356922292) [](ancestors = 356922292)

File: vnext/Microsoft.ReactNative.Cxx.UnitTests/JsonJSValueReader.cpp (hunk 21 above)
Comment (vmoroz) [reply vnext/Microsoft.ReactNative.Cxx.UnitTests/JsonJSValueReader.cpp]: Yes, someone already asked for it too. I will move it out to the Microsoft.ReactNative.Cxx in a future PR.

---
In reply to: [356922710](https://github.com/microsoft/react-native-windows/pull/3760#discussion_r356922710) [](ancestors = 356922710)

**Comment by NickGerleman** [comment]
I noticed that this seems to add a new directory structure to the project, both in terms of directory naming and organization. Is there a plan to move other future code to this structure? E.g. wondering why this wasn't done in something like ReactWindowsCore.

**Comment by vmoroz** [comment]
No, there is no new directory structure. At this point we are working on top of the Micah's prototype who created Microsoft.ReactNative project. At this point we just want to be consistent with what he started, but we already discussed renaming it to either React.Windows or ReactNative.Windows. Though, the question is how we can publish it as a NuGet package because Microsoft only owns the Microsoft top namespace and not the React or ReactNative namespace. We hope it will be fine to have ReactNative.Windows assemply while the NuGet packages has name that starts with Microsoft.

---
In reply to: [565261256](https://github.com/microsoft/react-native-windows/pull/3760#issuecomment-565261256) [](ancestors = 565261256)
